src/mappings/handlers.ts

```diff
@@ -0,0 +1,278 @@
+    let toChainId = BigInt.fromI32(7700) // todo
+    let pending = false
+
+    if (!swapSuccess) { // todo: verify how this reacts
+        receivedTokenValue = receivedTokenValue.minus(fee)
+    }
+
+    // update bridge txn
+    let txn = BridgeTransaction.load(kappa)
+    if (txn == null) {
+        txn = new BridgeTransaction(kappa)
+    }
+
+    if (txn.toTxnHash) {
+        pending = false
+    }
+
+    // update object
+    // txn.toTxnHash = null 
+    txn.kappa = kappa
+    txn.toTxnHash = txnHash.toHexString()
+
+    txn.toAddress = toAddress.toHexString()
+    
+    txn.receivedValue = receivedTokenValue
+
+    txn.toChainId = toChainId
+
+    txn.toChainBlock = blockNumber
+
+    txn.receivedTime = timestamp
+
+    txn.receivedTokenAddress = receivedTokenAddress.toHexString()
+    txn.receivedTokenSymbol = receivedTokenSymbol
+
+    txn.pending = pending
+
+    // save
+    txn.save()
+}
+
+export function handleTokenWithdrawAndRemove(event: TokenWithdrawAndRemove): void {
+    const blockNumber = event.block.number
+    const timestamp = event.block.timestamp
+    const txnHash = event.transaction.hash
+    const kappa = crypto.keccak256(ByteArray.fromHexString(txnHash.toHexString())).toHexString()
+
+    let toAddress = event.params.to
+    let fee = event.params.fee
+    let swapSuccess = event.params.swapSuccess
+    let tokenIndexTo = event.params.swapTokenIndex
+    let receivedTokenAddress = event.params.token
+    let receivedTokenSymbol = TokenDefinition.fromAddress(receivedTokenAddress).symbol
+    let receivedTokenValue = event.params.amount
+    let toChainId = BigInt.fromI32(7700) // todo
+    let pending = false
+
+    if (!swapSuccess) {
+        receivedTokenValue = receivedTokenValue.minus(fee)
+    }
+
+    // update bridge txn
+    let txn = BridgeTransaction.load(kappa)
+    if (txn == null) {
+        txn = new BridgeTransaction(kappa)
+    }
+
+    if (txn.toTxnHash) {
+        pending = false
+    }
+
+    // update object
+    // txn.toTxnHash = null 
+    txn.kappa = kappa
+    txn.toTxnHash = txnHash.toHexString()
+
+    txn.toAddress = toAddress.toHexString()
+    
+    txn.receivedValue = receivedTokenValue
+
+    txn.toChainId = toChainId
+
+    txn.toChainBlock = blockNumber
+
+    txn.receivedTime = timestamp
+
+    txn.receivedTokenAddress = receivedTokenAddress.toHexString()
+    txn.receivedTokenSymbol = receivedTokenSymbol
+
+    txn.pending = pending
+    txn.swapSuccess = swapSuccess
+
+    // save
+    txn.save()
+}
+
+export function handleTokenWithdraw(event: TokenWithdraw): void {
+    const blockNumber = event.block.number
+    const timestamp = event.block.timestamp
+    const txnHash = event.transaction.hash
+    const kappa = crypto.keccak256(ByteArray.fromHexString(txnHash.toHexString())).toHexString()
+
+    let toAddress = event.params.to
+    let fee = event.params.fee
+    let swapSuccess = null
+    let receivedTokenAddress = event.params.token // todo: indexer has other logic
```

Comment:
did you try to understand what's happening in the indexer?
this logic is completely different from that